### ISSUE TITLE
ArrayIndexOutOfBoundsException in FastLineAndPointRenderer during intensive plot rendering

### ISSUE THREAD

**Issue by jdpigeon** [halfhp/androidplot#15]
I was delighted to find that the FastLineAndPointRenderer class had been added, as it increased the fps of a live sensor data plot in my app to a consistent 40-50fps. 

However, I've been noticing occasional flickering of the graph, accompanied by index out of bounds exceptions thrown in FastLineAndPointRenderer. I took a look at the code myself and couldn't figure out how to fix it, though.

There's two cases I've noticed that will trigger the flickering:
1. The thread that adds values to the data series rendered by the FastLineAndPointRenderer is set to run in background priority
2. Two sensor data plots are displayed at the same time.

Here's the relevant part of the error code:
```
E/com.androidplot.Plot: Exception while rendering Plot.
java.lang.ArrayIndexOutOfBoundsException: length=428; index=428
at com.androidplot.xy.FastLineAndPointRenderer.onRender(FastLineAndPointRenderer.java:73)
at com.androidplot.xy.FastLineAndPointRenderer.onRender(FastLineAndPointRenderer.java:32)
at com.androidplot.ui.SeriesRenderer.render(SeriesRenderer.java:59)
at com.androidplot.xy.XYGraphWidget.drawData(XYGraphWidget.java:760)
at com.androidplot.xy.XYGraphWidget.doOnDraw(XYGraphWidget.java:461)
```

**Comment by halfhp** [comment]
Thanks for the report! I'll take a look and see what I can find!  Also would you be able to provide a little more information about your implementation?  In particular:

1. Are you using a default implementation of XYSeries (like SimpleXYSeries) or a custom implementation?
2. At what frequency is the sensor data coming in and modifying your series data?
3. Do you have synchronization in place to prevent the sensor updates from modifying the series data in the middle of a render loop?

Generally speaking, this class of issue tends be be caused by missing synchronization outside the Androidplot library.  Most likely the root cause is a race condition that has always existed and is only now being exposed due to the significant decrease in time spent in the render loop.

The most common cause of this issue is adding or removing an element to an XYSeries while Androidplot is in the middle of rendering it.  If you're using SimpleXYSeries you get some protection against that scenario for free via it's implementation of the PlotListener interface. 

Another way around this kind of issue is to use a fixed size series, such as the DemoApp's ring buffer implementation named ECGModel or FixedSizeEditableSeries in the upcoming 1.3 release.  It's worth noting however that while this approach avoids the ArrayIndexOutOfBoundsException, it's likely that at the point where the exception would have occurred will result in a sample in one of your render loops being off.  (probably not a concern when you're running @ 40+ fps)

**Comment by jdpigeon** [comment]
Hey halfhp!

That totally makes sense. Here's some more info about what caused the error:
1. I'm using a custom implementation of XYSeries which simply adds data points with LinkedList.addLast()
2. The sensor data's coming in fast, between 100-250hz
3. I don't have any synchronization setup between modifying and drawing the data. There's a seperate Thread and runnable for each action, with the draw thread running at 30hz and the data addition around 220hz.

Based on your feedback, I fixed the issue by handling synchronization with the PlotListener interface. My graphs run smoothly now without any noticeable change in performance.

For anyone else who has this problem in the future, what solved the flickering was to make my custom XYSeries class more like the SimpleXYSeries class in the library. Specifically, I implemented PlotListener, added ReentrantReadWriteLocks around my addLast and removeFirst calls and copied onBeforeDraw and onAfterDraw methods.

Thanks!
